File: backend/app/models/problem.py
from sqlalchemy import Column, Integer, String, Text, Enum as SQLEnum
from sqlalchemy.orm import relationship
import enum
from app.database import Base


class Difficulty(enum.Enum):
    Easy = "Easy"
    Medium = "Medium"
    Hard = "Hard"


class Problem(Base):
    __tablename__ = "problems"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    difficulty = Column(SQLEnum(Difficulty), nullable=False)
    description = Column(Text, nullable=False)
    time_limit = Column(Integer, default=2, nullable=False)  # seconds
    memory_limit = Column(Integer, default=256, nullable=False)  # MB

    # Relationships
    submissions = relationship("Submission", back_populates="problem")

File: springboot-backend/src/main/java/com/skillnest/entity/Problem.java
package com.skillnest.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "problems", indexes = {
    @Index(name = "idx_difficulty", columnList = "difficulty"),
    @Index(name = "idx_title", columnList = "title")
})
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Problem {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @NotBlank(message = "Title cannot be blank")
    @Column(nullable = false, unique = true, length = 255)
    private String title;
    
    @NotBlank(message = "Description cannot be blank")
    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;
    
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Difficulty difficulty;
    
    @Column(nullable = false)
    @Builder.Default
    private Integer timeLimit = 5; // seconds
    
    @Column(nullable = false)
    @Builder.Default
    private Integer memoryLimit = 256; // MB
    
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TestCase> testCases;
    
    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Submission> submissions;
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
}
